File: scaffold-app/src/main/java/com/example/scaffold/config/ShiroConfig.java
package com.example.scaffold.config;

import com.example.scaffold.repository.UserRepository;
import com.example.scaffold.security.JpaRealm;
import com.example.scaffold.security.ShiroAuthFilter;
import com.example.scaffold.service.PasswordService;
import org.apache.shiro.SecurityUtils;
import org.apache.shiro.authc.credential.HashedCredentialsMatcher;
import org.apache.shiro.crypto.hash.Sha256Hash;
import org.apache.shiro.mgt.DefaultSecurityManager;
import org.apache.shiro.realm.Realm;
import org.apache.shiro.session.mgt.DefaultSessionManager;
import org.apache.shiro.session.mgt.eis.MemorySessionDAO;
import org.springframework.boot.web.servlet.FilterRegistrationBean;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.core.Ordered;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Apache Shiro configuration.
 * <p>
 * Integrates Shiro's authentication/authorization core (Realm, CredentialsMatcher,
 * SecurityManager, SessionManager) with Spring Boot 3. A custom Jakarta filter
 * ({@link ShiroAuthFilter}) enforces the URL filter chain, since the official Shiro
 * Spring Boot web starter is javax-based and incompatible with Spring Boot 3.
 * <p>
 * The filter chain explicitly lists every endpoint so that none is missed:
 * <ul>
 *   <li>{@code /api/login}, {@code /api/logout} &mdash; anonymous (login/logout endpoints)</li>
 *   <li>{@code /api/version}, {@code /api/ping} &mdash; anonymous (health/version checks)</li>
 *   <li>{@code /error} &mdash; anonymous (Spring error path)</li>
 *   <li>{@code /**} &mdash; authenticated (covers /api/hello, /api/users, /api/me and anything else)</li>
 * </ul>
 */
@Configuration
public class ShiroConfig {

    private static final long SESSION_TIMEOUT_MILLIS = 30 * 60 * 1000L;

    @Bean
    public HashedCredentialsMatcher credentialsMatcher() {
        HashedCredentialsMatcher matcher = new HashedCredentialsMatcher();
        matcher.setHashAlgorithmName(Sha256Hash.ALGORITHM_NAME);
        matcher.setHashIterations(PasswordService.ITERATIONS);
        matcher.setStoredCredentialsHexEncoded(true);
        matcher.setHashSalted(true);
        return matcher;
    }

    @Bean
    public Realm realm(UserRepository userRepository, HashedCredentialsMatcher credentialsMatcher) {
        return new JpaRealm(userRepository, credentialsMatcher);
    }

    @Bean
    public DefaultSessionManager sessionManager() {
        DefaultSessionManager sessionManager = new DefaultSessionManager();
        sessionManager.setSessionDAO(new MemorySessionDAO());
        sessionManager.setGlobalSessionTimeout(SESSION_TIMEOUT_MILLIS);
        sessionManager.setDeleteInvalidSessions(true);
        return sessionManager;
    }

    @Bean
    public DefaultSecurityManager securityManager(Realm realm, DefaultSessionManager sessionManager) {
        DefaultSecurityManager securityManager = new DefaultSecurityManager(realm);
        securityManager.setSessionManager(sessionManager);
        // Make the security manager available to SecurityUtils.getSubject() / SecurityUtils.getSecurityManager().
        SecurityUtils.setSecurityManager(securityManager);
        return securityManager;
    }

    @Bean
    public ShiroAuthFilter shiroAuthFilter(DefaultSecurityManager securityManager) {
        return new ShiroAuthFilter(securityManager, filterChain());
    }

    @Bean
    public FilterRegistrationBean<ShiroAuthFilter> shiroAuthFilterRegistration(ShiroAuthFilter shiroAuthFilter) {
        FilterRegistrationBean<ShiroAuthFilter> registration = new FilterRegistrationBean<>(shiroAuthFilter);
        registration.addUrlPatterns("/*");
        registration.setOrder(Ordered.HIGHEST_PRECEDENCE + 10);
        registration.setName("shiroAuthFilter");
        return registration;
    }

    /**
     * Ordered filter chain: anon paths first, then a catch-all requiring authentication.
     * Every controller endpoint is covered either explicitly or by the catch-all.
     */
    private Map<String, String> filterChain() {
        Map<String, String> chain = new LinkedHashMap<>();
        chain.put("/api/login", "anon");
        chain.put("/api/logout", "anon");
        chain.put("/api/version", "anon");
        chain.put("/api/ping", "anon");
        chain.put("/error", "anon");
        chain.put("/**", "authc");
        return chain;
    }
}

File: python/app/security/shiro_config.py
"""Shiro 过滤链配置。

对应原项目 ``config/ShiroConfig.java`` 的 ``filterChain()``:
显式列出每个端点的匿名(anon)/ 需认证(authc)规则,顺序匹配,首条命中生效,
未命中项默认 ``authc``。保证没有任何端点被意外漏配。

静态页面(/login.html 等)设为 anon,页面自身通过 /api/me 判断登录态。
"""
from __future__ import annotations

from collections import OrderedDict

#: Cookie 名,对应 ``ShiroAuthFilter.SESSION_COOKIE_NAME``。
SESSION_COOKIE_NAME = "SHIRO_SESSION_ID"

ACTION_ANON = "anon"
ACTION_AUTHC = "authc"


def filter_chain() -> "OrderedDict[str, str]":
    """返回有序过滤链(模式 -> 动作),与 Java ``filterChain()`` 逐条一致。"""
    chain: "OrderedDict[str, str]" = OrderedDict()
    chain["/api/login"] = ACTION_ANON
    chain["/api/logout"] = ACTION_ANON
    chain["/api/register"] = ACTION_ANON
    chain["/api/captcha"] = ACTION_ANON
    chain["/api/version"] = ACTION_ANON
    chain["/api/ping"] = ACTION_ANON
    chain["/login.html"] = ACTION_ANON
    chain["/register.html"] = ACTION_ANON
    chain["/index.html"] = ACTION_ANON
    chain["/"] = ACTION_ANON
    chain["/error"] = ACTION_ANON
    chain["/**"] = ACTION_AUTHC
    return chain


def matches(pattern: str, path: str) -> bool:
    """Ant 风格匹配,对应 ``ShiroAuthFilter.matches``。

    支持 ``/**``(任意深度)、``/*``(单段)与精确匹配。
    """
    if pattern == "/**":
        return True
    if pattern.endswith("/**"):
        prefix = pattern[: -len("/**")]
        return path.startswith(prefix)
    if pattern.endswith("/*"):
        prefix = pattern[: -len("/*")]
        if not path.startswith(prefix):
            return False
        remainder = path[len(prefix):]
        return "/" not in remainder
    return pattern == path


def match_action(path: str) -> str:
    """返回路径命中的动作;未命中则安全默认为 ``authc``。"""
    for pattern, action in filter_chain().items():
        if matches(pattern, path):
            return action
    return ACTION_AUTHC
